M1FExplained/Chapter10/Exercises/Exercise02.lean

```diff
@@ -2,14 +2,117 @@ import Mathlib.Tactic
 
 namespace Chapter10.Exercise02
 
+-- part a
+
+/-
+Show that if a, b are positive integers and d = gcd(a, b), there exists positive integers s, t such
+that d = s * a - t * b
+-/
+
+/-
+The key to the solution to the question is the following trick:
+
+s * a + t * b = (s + p * b) * a + (t - p * a) * b for any positive integer p.
+
+Thus we can make the coefficients of a, b as positive/negative as we like.
+-/
+
 /-
-Part a asks us to prove that for natural numbers a, b, ∃ s t : ℕ such that gcd(a, b) = s * a - t * b
+First of all, a few helper lemmas are proven. The first two deal with the nuances of integer division
+in Lean. The last two solve the bulk of the question, leaving only a small amount of work to be done 
+in the actual question statment.
 -/
 
-lemma part_a (a b d : ℕ) (hd : d = Nat.gcd a b) : ∃ (s t : ℕ),  d = s * a - t * b := by sorry
+lemma Int.sub_ediv_of_dvd_right {a b c : ℤ} (H : c ∣ b) : (a - b) / c = a / c - b / c := by
+  simp only [sub_eq_add_neg, Int.add_ediv_of_dvd_right <| Int.dvd_neg.2 H, Int.neg_ediv_of_dvd H]
+
+lemma Int.sub_lt_div_mul_self {a b : ℤ} (hb : 0 < b) : a - b < a / b * b := by
+  by_contra h
+  push_neg at h
+  have := Int.le_ediv_of_mul_le hb h
+  rw [Int.sub_ediv_of_dvd_right <| dvd_rfl, Int.ediv_self hb.ne'] at this
+  linarith
+
+/- 
+helper_1 says that given s * a + t * b = n, we can find new coefficients s', t' for a, b 
+with 0 < s' so the same equality holds.
+-/
+
+lemma helper_1 (a b n : ℤ) (hb : 0 < b) (hab : ∃ (s t : ℤ), s * a + t * b = n) : 
+    ∃ (s' t' : ℤ), 0 < s' ∧ s' * a + t' * b = n := by 
+  rcases hab with ⟨s, t, hst⟩
+  set p := (b - s) / b with hp
+  use s + p * b, t - p * a
+  constructor
+  · rw [hp, ← neg_lt_iff_pos_add']
+    calc
+      -s = (b - s) - b := by norm_num
+      _ < (b - s) / b * b := by exact Int.sub_lt_div_mul_self hb
+  · linear_combination hst
 
 /-
-Part b asks us to find such s and t for the examples in exercise 01, Lean provides us with these: 
+helper_2 builds on helper_1 but now adds the restriction that the new coefficient t' satisfies t' < 0.
+This effectivley solves the question, save for some changes of sign and the assumption that we can 
+initially find any s, t such that s * a + t * b = gcd(a, b).
+-/
+
+lemma helper_2 (a b n : ℤ) (ha : 0 < a) (hb : 0 < b) (hab : ∃ (s t : ℤ), s * a + t * b = n) : 
+    ∃ (s' t' : ℤ), 0 < s' ∧ t' < 0 ∧ s' * a + t' * b = n := by 
+    match helper_1 a b n hb hab with
+    |⟨s, t, ⟨h1, h2⟩⟩ => 
+    apply Or.elim (Classical.em (t < 0))
+    <;> intro ht
+    · use s, t
+      repeat (any_goals constructor)
```

Comment:
I get linter warnings telling me that there aren't any goals left at this point. Don't you?